### PR TITLE
[UI] Add timespan selection and auto refresh for metrics tab in node page

### LINKED ISSUE #2776: [UI New Node Page] Add metrics graph in Node Page
follows: #2765 

**Component**: ui, nodes

**What should be done**:
- [x] Add the graphs with the dafault timespan of 'Last 24hours'
- [x] the choices of timespan are last 24hours/ last 7 days/ last 1 hour ( by default is last 24 hours)
- [x] Update the URL with the query of current timespan
- [x] the graphs can be auto-refreshed

**Implementation proposal** (strongly recommended):
image
Add one more query param in the URL for the three timespan choices.

Last 7days: &from=now-7d
Last 24hours: &from=now-24h
Last 1 hour: &from=now-1h

**Test plan**:
TBD

### PR DESCRIPTION
**Pull request by ChengYanJin**
**Component**: ui, nodes

**Context**:  Add timespan selection and auto-refresh for metrics tab in the Node page

**Summary**:
- Add timespan selection in `metrics` tab, by default is Last 24 hours
- Add selected timespan in URL query
- Add auto-fresh of graphs every 1 min
- Keep the selected timespan when we switch back to metrics tab
- When we select a node from the list, we don't add the default the timespan(from=now-24h) in the URL. 
- The style of dropdown component is not the same as design. (Need to work on it in Core-ui)
- Launch a long-time running saga at the root of saga to watch the Refresh_NODESTATS action.

**Acceptance criteria**: 
image

image

---

Closes: #2776

### CODE REVIEW

File: ui/src/components/NodePageMetricsTab.js [hunk 1]
```diff
@@ -212,8 +239,48 @@ const NodePageMetricsTab = (props) => {
   };
   const lineConfig = { strokeWidth: 1.5 };
 
+  const metricsTimeSpanItems = [
+    {
+      label: LAST_SEVEN_DAYS,
+      onClick: () => {
+        dispatch(updateNodeStatsAction({ metricsTimeSpan: LAST_SEVEN_DAYS }));
+        updateMetricsGraph();
```
Comment (gdemonet) [inline ui/src/components/NodePageMetricsTab.js]: Really more a question than a suggestion, but wouldn't it be "cleaner" to handle this at the saga level?

Instead of a single "refresh" saga which is going to send this "update metrics" action periodically, we could have it use a `race` effect, between `take(TIMESPAN_UPDATED)` and `delay(REFRESH_PERIOD)`, making sure we won't have too many call sites for the chart update... WDYT?

File: ui/src/components/NodeListTable.js [hunk 2]
```diff
@@ -390,8 +412,12 @@ const NodeListTable = (props) => {
       location.pathname.endsWith('volumes') ||
       location.pathname.endsWith('pods');
 
-    if (isTabSelected) {
-      // When switch between the nodes, keep the same tab selected
+    if (isTabSelected && location.pathname.endsWith('metrics')) {
+      // Keep the same timespan selected for metrics tab, when switch between the nodes
+      const currentTab = location?.pathname?.split('/')?.pop();
+      history.push(`/newNodes/${nodeName}/${currentTab}?from=${queryTimeSpan}`);
```
Comment (gdemonet) [inline ui/src/components/NodeListTable.js]: This means we loose the other query parameters (like the table filters)... But I see it's not just related to this change, because we had this bug before apparently.

Wouldn't it be simpler to just keep the exact same URL and switch only the Node name part?
Something like:

```javascript
const newPath = location.pathname.replace(/\/newNodes\/[^\/]*\//, `/newNodes/${nodeName}/`);
history.push(newPath);
```

If you really need to sanitize the query string, I think you'd be able to by using `URLSearchParams`.

File: ui/src/components/NodePageMetricsTab.js [hunk 3]
```diff
@@ -245,6 +251,9 @@ const NodePageMetricsTab = (props) => {
       onClick: () => {
         dispatch(updateNodeStatsAction({ metricsTimeSpan: LAST_SEVEN_DAYS }));
         updateMetricsGraph();
+        history.push(
+          `/newNodes/${selectedNodeName}/metrics?from=${QUERY_LAST_SEVEN_DAYS}`,
+        );
```
Comment (gdemonet) [inline ui/src/components/NodePageMetricsTab.js]: Ditto, this approach cleans off the other parts of the querystring, which is not desired behaviour IMO

File: ui/src/containers/NodePageRSP.js [hunk 4]
```diff
@@ -53,7 +56,7 @@ const NodePageRSP = (props) => {
   // we should initialize the `metricsTimeSpan` in redux base on the URL query.
   const query = useQuery();
   const queryTimespan = query.get('from');
-  let metricsTimeSpan;
+  let metricsTimeSpan = LAST_TWENTY_FOUR_HOURS;
```
Comment (gdemonet) [inline ui/src/containers/NodePageRSP.js]: Why this change? The `default` case below seems sufficient

File: ui/src/ducks/app/monitoring.js [hunk 5]
```diff
@@ -87,6 +89,9 @@ const defaultState = {
   },
   nodeStats: {
     metricsTimeSpan: LAST_TWENTY_FOUR_HOURS,
+    instanceIP: '',
+    controlPlaneInterface: '',
+    workloadPlaneInterface: '',
```
Comment (gdemonet) [inline ui/src/ducks/app/monitoring.js:109]: I think it makes sense to maintain these values in the state (in essence, this represents the "selected node" for which we'll pull metrics). However, doing this should allow us to only set them once, when we mount the metrics tab component (because we know which Node is selected), and never touch them again.

In practice, it means that:

- when mounting / rendering the component, you dispatch this once:

```javascript
      updateNodeStatsAction({
        metricsTimeSpan,
        instanceIP,
        controlPlaneInterface,
        workloadPlaneInterface,
      })
```

- the fetch saga reads from the state this information about the selected node and timespan, not from the payload, so you dispatch this at render, after the previous one:

```javascript
    fetchNodeStatsAction()
```

- the refresh saga doesn't care about the state, it's only periodically sending a `FETCH_NODE_STATS` action until it receives a halt (see later comment on how to write the refresh saga)

File: ui/src/ducks/app/monitoring.js [hunk 6]
```diff
@@ -725,6 +743,33 @@ export function* fetchNodeStats({ payload }) {
...
+      },
+    });
+  }
+  yield delay(REFRESH_METRCIS_GRAPH);
+
+  const isRefreshing = yield select(isNodeStatsRefreshing);
+  if (isRefreshing) {
+    yield call(refreshNodeStats);
+  }
+}
```
Comment (gdemonet) [inline ui/src/ducks/app/monitoring.js]: This pattern for refresh sagas was already discussed with Carlito, back in the days, and we came to the conclusion that it had big drawbacks (you can loose a "STOP" action, or trigger the refresh loop multiple times in parallel, and overall, it's not easy to read). See #2035.

In practice, with the suggestion from a previous comment, that would look like:

```javascript
export function* refreshNodeStats() {
  while (true) {
    yield take(REFRESH_NODESTATS);
    while (true) {
      yield fork(fetchNodeStats);
      const {interrupt, update} = yield race({
        interrupt: take(STOP_REFRESH_NODESTATS),
        // If the refresh period expires before we receive a halt,
        // we can refresh the stats
        requeue: delay(DELAY_TIMEOUT),
        // NOTE: This action doesn't exist, but it should, so that
        // whenever we change one of the parameters for "fetchNodeStats",
        // it gets triggered again
        update: take(UPDATE_SELECTED_NODE),
      });
      if (interrupt) {
        yield cancel(fetchNodeStats);
        break;
      }
    }
  }
}
```

File: ui/src/containers/NodePageRSP.js [hunk 7]
```diff
@@ -53,19 +54,38 @@ const NodePageRSP = (props) => {
...
+      case LAST_SEVEN_DAYS:
+        queryTimespan = QUERY_LAST_SEVEN_DAYS;
+        break;
+      case LAST_ONE_HOUR:
+        queryTimespan = QUERY_LAST_ONE_HOUR;
+        break;
+      default:
+        queryTimespan = QUERY_LAST_TWENTY_FOUR_HOURS;
+    }
   }
```
Comment (gdemonet) [inline ui/src/containers/NodePageRSP.js]: Would be nicer to use maps, especially if we start adding more options.

**Review by gdemonet — CHANGES_REQUESTED** [review]
Just some code review (mostly Sagas again :wink:), I'll test it soon!

File: ui/src/ducks/app/monitoring.js (hunk 6 above)
Comment (ChengYanJin) [reply ui/src/ducks/app/monitoring.js]: hello @gdemonet ,

I wanted to address this dept ticket for a long time! Let's work on it now!

I have one questions in my mind:
What does `yield take(REFRESH_NODESTATS);` do? In this case, I think we may never have the `REFRESH_NODESTATS` dispatched.

Comment (gdemonet) [reply ui/src/ducks/app/monitoring.js]: `take(SOME_EVENT)` is blocking until `SOME_EVENT` is seen.

So, the idea is to replace the `takeFirst` effect creator:

```javascript
while (true) {
  // wait for the start event
  yield take("START_REFRESH");
  // once we've seen this start event, we can enter the refresh loop
  while (true) {
    // trigger the data update
    yield fork(fetchData);
    // wait for the stop event for at most the refresh period
    const {interrupt} = yield race({
      interrupt: take("STOP_REFRESH"),
      requeue: delay(REFRESH_PERIOD),
    });
    if (interrupt) {
      // we received the stop event, so we get out of the refresh loop
      // and we will wait for the next start event
      break;
    }
  }
}
```

File: ui/src/containers/NodePageRSP.js (hunk 4 above)
Comment (ChengYanJin) [reply ui/src/containers/NodePageRSP.js]: Actually, and we set the default value in the saga state. 
Maybe it doesn't hurt to set again here. 

``` javascript
  const query = useQuery();
  const nodeMetricsTimeSpan = useSelector(
    (state) => state.app.monitoring.nodeStats.metricsTimeSpan,
  );

  let metricsTimeSpan;
  const queryTimespan = query.get('from');

  if (queryTimespan) {
    // If timespan query specified in query string
    metricsTimeSpan = queryTimeSpansCodes?.find(
      (timespan) => timespan.label === queryTimespan,
    )?.value;
  } else {
    metricsTimeSpan = nodeMetricsTimeSpan;
  }
```


```javascript
nodeStats: {
    metricsTimeSpan: LAST_TWENTY_FOUR_HOURS,
    ...
}
```

File: ui/src/components/NodePageMetricsTab.js (hunk 1 above)
Comment (ChengYanJin) [reply ui/src/components/NodePageMetricsTab.js]: Definitely right! 
Will dispatch `updateNodeStatsFetchArgumentAction` to the `race` everytime we change the timespan or node(information)

File: ui/src/components/NodeListTable.js [hunk 8]
```diff
@@ -391,12 +390,14 @@ const NodeListTable = (props) => {
       location.pathname.endsWith('pods');
 
     if (isTabSelected) {
-      // When switch between the nodes, keep the same tab selected
-      const currentTab = location?.pathname?.split('/')?.pop();
-      history.push(`/newNodes/${nodeName}/${currentTab}`);
+      const newPath = location.pathname.replace(
+        // eslint-disable-next-line no-useless-escape
```
Comment (gdemonet) [inline ui/src/components/NodeListTable.js]: If there is a useless escape, maybe we should fix it?

File: ui/src/components/NodeListTable.js [hunk 9]
```diff
@@ -391,12 +390,14 @@ const NodeListTable = (props) => {
...
+        /\/newNodes\/[^\/]*\//,
+        `/newNodes/${nodeName}/`,
+      );
+      history.push(newPath);
     } else {
-      // Set Health tab as default tab
-      history.push(`/newNodes/${nodeName}/health`);
+      history.push({ pathname: `/newNodes/${nodeName}/health` });
```
Comment (gdemonet) [inline ui/src/components/NodeListTable.js]: We'll loose URL search params here as well :wink: 
In any case, I think this URL parsing in the component is not the right approach, we should instead rely on react-router's `Route` matching, that'd be much cleaner and safer. Let's open a debt ticket for this, we'll try to refactor other pages at the same time.

File: ui/src/containers/NodePageMetricsTab.js [hunk 10]
```diff
@@ -57,11 +68,22 @@ const LoaderContainer = styled(Loader)`
...
+  justify-content: flex-end;
+  padding-right: ${padding.base};
+`;
+
 const NodePageMetricsTab = (props) => {
   const { nodeStats } = props;
+  const dispatch = useDispatch();
   const theme = useSelector((state) => state.config.theme);
+  const history = useHistory();
+  const query = new URLSearchParams(history?.location?.search);
```
Comment (gdemonet) [inline ui/src/containers/NodePageMetricsTab.js]: Is there a difference between this and the `useQuery` hook? Why do we need to rely on `useHistory().location` rather than `useLocation()` directly?

File: ui/src/containers/NodePageMetricsTab.js [hunk 11]
```diff
@@ -212,8 +236,65 @@ const NodePageMetricsTab = (props) => {
...
+    }
+  };
+
+  const metricsTimeSpanItems = [
+    {
+      label: LAST_SEVEN_DAYS,
+      onClick: () => {
+        dispatch(
+          updateNodeStatsAction({ metricsTimeSpan: LAST_SEVEN_DAYS }),
+          updateNodeStatsFetchArgumentAction(),
```
Comment (gdemonet) [inline ui/src/containers/NodePageMetricsTab.js]: Would be a bit cleaner to differenciate between the two:

- `UPDATE_NODESTATS` is used to update `app.monitoring.nodeStats.metrics`
- `UPDATE_NODESTATS_FETCH_ARG` is used to update the other keys used as fetch arguments

This way, we could:

- only dispatch one action, instead of two
- never forget to send the `UPDATE_FETCH_ARG` action when needed, because that's the only way to actually update them
- rely on the main `UPDATE_NODESTATS` if we had other sagas that need to react on some metrics update

File: ui/src/containers/NodePageMetricsTab.js [hunk 12]
```diff
@@ -212,8 +236,65 @@ const NodePageMetricsTab = (props) => {
...
+      onClick: () => {
+        dispatch(
+          updateNodeStatsAction({ metricsTimeSpan: LAST_ONE_HOUR }),
+          updateNodeStatsFetchArgumentAction(),
+        );
+        writeUrlTimeSpan(LAST_ONE_HOUR);
+      },
+      selected: metricsTimeSpan === LAST_ONE_HOUR,
+    },
+  ];
```
Comment (gdemonet) [inline ui/src/containers/NodePageMetricsTab.js]: Could be refactored:

```suggestion
  const metricsTimeSpanItems = [
    LAST_SEVEN_DAYS, LAST_TWENTY_FOUR_HOURS, LAST_ONE_HOUR
  ].map(
    option => ({
      label: option,
      onClick: () => {
        dispatch(
          updateNodeStatsFetchArgumentAction({ metricsTimeSpan: option }),
        );
        writeUrlTimeSpan(option);
      },
      selected: metricsTimeSpan === option,
    })
  );
```

File: ui/src/containers/NodePageMetricsTab.js [hunk 13]
```diff
@@ -212,8 +236,65 @@ const NodePageMetricsTab = (props) => {
...
+
+  const metricsTimeSpanDropdownItems = metricsTimeSpanItems?.filter(
+    (mTS) => mTS.label !== metricsTimeSpan,
+  );
```
Comment (gdemonet) [inline ui/src/containers/NodePageMetricsTab.js:263]: I'm surprised, does it mean we only have one item in the dropdown?

**Review by gdemonet — CHANGES_REQUESTED** [review]
Mostly a change on how we handle updating the "fetch args"

File: ui/src/components/NodeListTable.js (hunk 9 above)
Comment (gdemonet) [reply ui/src/components/NodeListTable.js]: #2838

File: ui/src/containers/NodePageMetricsTab.js (hunk 13 above)
Comment (ChengYanJin) [reply ui/src/containers/NodePageMetricsTab.js:263]: Nope. The items we have in the dropdown are all the options **exclude** the current selected`metricsTimeSpan`

File: ui/src/components/NodeListTable.js (hunk 9 above)
Comment (ChengYanJin) [reply ui/src/components/NodeListTable.js]: You are right! Thanks for creating the debt ticket!
We're gonna lose the search params if we want to add filters for the table.

File: ui/src/components/NodeListTable.js (hunk 8 above)
Comment (ChengYanJin) [reply ui/src/components/NodeListTable.js]: It's actually an [eslint-error](https://stackoverflow.com/a/47277197) because of the regex: Unnecessary escape character: \/ no-useless-escape 
But I think the regex is correct. That's why I disable it, maybe the message is not easy to understand.

Comment (slaperche-scality) [reply ui/src/components/NodeListTable.js]: As mentioned in the answer you linked, you don't need to escape the forward slash inside the character class e.g. you could write `/\/newNodes\/[^/]*\//` and have the same behavior:

```
> var nodeName = "foo";
undefined
> var str = "/newNodes/BAR/";
undefined
> str.replace(/\/newNodes\/[^\/]*\//, `/newNodes/${nodeName}/`); 
'/newNodes/foo/'
> str.replace(/\/newNodes\/[^/]*\//, `/newNodes/${nodeName}/`); 
'/newNodes/foo/'
```

Did you try that version?
Is ESLint happy with this regex? Or do you still have the warning?

Comment (ChengYanJin) [reply ui/src/components/NodeListTable.js]: I confirm `/\/newNodes\/[^/]*\//` is working correctly!

File: ui/src/containers/NodePageMetricsTab.js (hunk 13 above)
Comment (gdemonet) [reply ui/src/containers/NodePageMetricsTab.js:263]: :man_facepalming: Oopsie, I read that too fast! Thanks for the translation :)

**Review by gdemonet — APPROVED** [review]
LGTM!
